### PR TITLE
Add error reason for fwrite in sys.save

### LINKED ISSUE #8856: Output errno when fwrite fails in Sys_Save (and errno is supported)

### PR DESCRIPTION
**Pull request by Jhonnyg**
Fixed an issue where sys.save doesn't output the reason why it couldn't write to a file.

Fixes #8856

### CODE REVIEW

File: engine/script/src/script_sys.cpp [hunk 1]
```diff
@@ -161,13 +161,14 @@ union SaveLoadBuffer
...
-                return luaL_error(L, "Could not open the file %s", tmp_filename);
-            #endif
+
+        #if defined(DM_NO_ERRNO)
+            return luaL_error(L, "Could not open the file %s", tmp_filename);
+        #else
+            char errmsg[128] = {};
+            dmStrError(errmsg, sizeof(errmsg), errno);
+            return luaL_error(L, "Could not open the file %s, reason: %s.", tmp_filename, errmsg);
+        #endif
```
Comment (Jhonnyg) [inline engine/script/src/script_sys.cpp:171]: I flipped the double negativeness of the ifdef

File: engine/script/src/script_sys.cpp [hunk 2]
```diff
@@ -177,7 +178,14 @@ union SaveLoadBuffer
         if (!result)
         {
             dmSys::Unlink(tmp_filename);
+
+        #if defined(DM_NO_ERRNO)
             return luaL_error(L, "Could not write to the file %s.", filename);
+        #else
+            char errmsg[128] = {};
+            dmStrError(errmsg, sizeof(errmsg), errno);
+            return luaL_error(L, "Could not write to the file %s, reason: %s.", filename, errmsg);
```
Comment (Jhonnyg) [inline engine/script/src/script_sys.cpp:187]: According to unix and windows manuals, fwrite _should_ set errno when it fails, so this should be enough..

File: engine/script/src/script_sys.cpp (hunk 1 above)
Comment (JCash) [reply engine/script/src/script_sys.cpp:171]: `dmStrError()` does the same thing when DM_NO_ERRNO is defined (See dstrings.cpp:236)
That way you don't have to use another `ifdef`
I would recommend removing the "if defined(DM_NO_ERRNO)"

**Review by JCash — CHANGES_REQUESTED** [review]
Approved, but with some minor change request.

File: engine/script/src/script_sys.cpp (hunk 1 above)
Comment (Jhonnyg) [reply engine/script/src/script_sys.cpp:171]: Not sure I understand, in this case and after fopen a few lines above, we want to pass in the errno into the function to actually parse it, and if I remove the ifdef errno would be undefined for the platforms where it isn't supported, right?

**Pull request merged**